python/cudf/dataframe/multiindex.py

```diff
@@ -0,0 +1,341 @@
+        self.names = names
+
+    def _validate_levels_and_codes(self, levels, codes):
+        levels = np.array(levels)
+        if len(levels) != len(codes.columns):
+            raise ValueError('MultiIndex has unequal number of levels and '
+                             'codes and is inconsistent!')
+        code_length = len(codes[codes.columns[0]])
+        for index, code in enumerate(codes):
+            if code_length != len(codes[code]):
+                raise ValueError('MultiIndex length of codes does not match '
+                                 'and is inconsistent!')
+        for index, code in enumerate(codes):
+            if codes[code].max() > len(levels[index])-1:
+                raise ValueError('MultiIndex code %d contains value %d larger '
+                                 'than maximum level size at this position')
+
+    def copy(self, deep=True):
+        if(deep):
+            result = deepcopy(self)
+        else:
+            result = copy(self)
+        result.name = self.name
+        return result
+
+    def _popn(self, n):
+        """ Returns a copy of this index without the left-most n values.
+
+        Removes n names, labels, and codes in order to build a new index
+        for results.
+        """
+        from cudf import DataFrame
+        codes = DataFrame()
+        for idx in self.codes.columns[n:]:
+            codes.add_column(idx, self.codes[idx])
+        result = MultiIndex(self.levels[n:], codes)
+        result.names = self.names[n:]
+        return result
+
+    def __repr__(self):
+        return "MultiIndex(levels=" + str(self.levels) +\
+               ",\ncodes=" + str(self.codes) + ")"
+
+    @property
+    def labels(self):
+        warnings.warn("This feature is deprecated in pandas and will be"
+                      "dropped from cudf as well.", FutureWarning)
+        return self.codes
+
+    def _compute_validity_mask(self, df, row_tuple):
+        """ Computes the valid set of indices of values in the lookup
+        """
+        validity_mask = []
+        for i, element in enumerate(row_tuple):
+            index_of_code_at_level = None
+            for level_index in range(len(self.levels[i])):
+                if self.levels[i][level_index] == element:
+                    index_of_code_at_level = level_index
+                    break
+            if index_of_code_at_level is None:
+                raise KeyError(element)
+            matches = []
+            for k, code in enumerate(self.codes[self.codes.columns[i]]):
+                if k in validity_mask or len(validity_mask) == 0:
+                    if code == index_of_code_at_level:
+                        matches.append(k)
+            # matches = self.codes.map([j if c[i] == index_of_code_at_level\
+            #       else None for j,c in codes])
+            if len(matches) != 0:
+                validity_mask = matches
+        return validity_mask
+
+    def get_row_major(self, df, row_tuple):
```

Comment:
Shouldn't this be `_get_row_major`?

Reply:
certainly!

Reply:
I don't think so, since it is called by the dataframe `getitem` and `loc` methods. Perhaps a better abstraction might be to just call a `get` which determines whether or not to do _get_row or _get_column. I'm not sure that refactor is worth it at this late hour. :)

Reply:
If `MultiIndex` is an object that we expect users to interact with directly, then we should limit the methods that it makes publicly visible. Here are the methods users can see:

```
'argsort',
 'copy',
 'deserialize',
 'equals',
 'find_segments',
 'from_frame',
 'from_pandas',
 'from_product',
 'from_tuples',
 'get_column_major',
 'get_row_major',
 'join',
 'max',
 'min',
 'rename',
 'serialize',
 'sum',
 'take',
 'to_arrow',
 'to_dlpack',
 'to_pandas'
```

If `get_row_major` and `get_column_major` are only to be used internally by cudf, we should not make them non-public.